Most Profitable Free Apps For iPhone and Android

As the company only develops free apps, profit is made through in-app ads which is based on the number of installs. The aim of this project is to provide the company's developers with data that will give them a better of idea of what type of app to develop. 

-We firstly import the reader fuction from CSV
-Then use the built function to open the file
-After we use the list function to convert the read_file to a list of lists 
-Finally we split the first row and the rest of the data because it contains the header names


In [1]:
from csv import reader
#The Apple Store Data#
open_file = open('AppleStore.csv')
read_file = reader(open_file)
ios_file = list(read_file)
ios_header = list(ios_file[0])
ios_data = list(ios_file[1:])

#The Android Store Data#
open_file = open('googleplaystore.csv')
read_file = reader(open_file)
android_file = list(read_file)
android_header = android_file[0]
android_data = android_file[1:]


In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line between rows
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

print(android_header)
print('\n')
explore_data(android_data, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


From the above some columns that might help us with the analysis are: 'Ratings', 'Reviews', 'Intalls', 'Price' and 'Genres'. We can also see from here the Android data countains 10,841 rows and 13 colums. 

In [3]:
print(ios_header)
print('\n')
explore_data(ios_data, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


For the Apple Store, the most interesting headings seem to be 'track_name', 'price', 'rating_count_tot', 'user_rating', and 'prime_genre'. Further description of headings can be found here: https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home


In the discussion section of the Google Play data set, it is pointed out that row 10472 has an error. Printing the header and the incorrect row and a correct row to compare if it is indeed incorrect. 

In [4]:
print(android_header)  # header
print('\n')
print(android_data[10472])  # incorrect row
print('\n')
print(android_data[1])      # correct row

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


We can see user rating for the app 'Life Made WI-Fi' is '19', the maximum rating in the Google Play store is 5. As we do not know the correct rating for this row, we will delete this row/outlier from the data. 

In [5]:
del android_data[10472]

In [6]:
len(android_data)

10840

In [7]:
duplicate_apps = []
unique_apps= []

for app in android_data:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

print('No. of duplicate apps:', len(duplicate_apps))

No. of duplicate apps: 1181


From the above we can see that there are some duplicate apps

In [9]:
print('Some Examples of Duplicate Apps:', duplicate_apps[:15])

Some Examples of Duplicate Apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


In [10]:
for app in android_data:
    app_name = app[0]
    if app_name == 'Quick PDF Scanner + OCR FREE':
        print(app)

['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80804', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']


We can see that 'Quick PDF Scanner + OCR FREE' is duplicated three times. The main difference is for fourth entry, which is the number of reviews. We do not want to have duplicate apps in our data, so we will create a criteria for which line to keep. It makes sense to keep the line with the highest number of reviews, this gives a better reflection of the rating of the app.

We need to map the app to highest number of reviews, for that we will create a dictionary. 

In [11]:
reviews_max = {}
for app in android_data:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

print(len(reviews_max))

9659


We previously saw that the number of duplicate apps was 1181, therefore the difference between the total number apps and duplicate apps should be the same as the number of unique keys/apps in the dictionary.

In [12]:
print(len(android_data) - len(duplicate_apps))
print(len(reviews_max))

9659
9659


In [13]:
android_clean = []
already_added = []

for app in android_data:
    name = app[0]
    no_of_reviews = float(app[3])
    if reviews_max[name] == no_of_reviews and name not in already_added:
        android_clean.append(app)
        already_added.append(name)
        
print(len(android_clean))

9659


We needed to create a list of clean apps, so we create a condition to check if the app reviews is the same as the reviews in reviews_max and if it's not in the already_added list. 

Apple store data has no duplicates.

Some apps are not in english, we will remove these from the data. English letters correspond to numbers from 0 to 127 according to ASCII. We will build a function to check if any letters in a string are greater than 127.

In [14]:
def in_english(string):
    for character in string:
        if ord(character) > 127:
            return False
    return True
    
print(in_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(in_english('Instagram'))
print(in_english('Instachat 😜'))

False
True
False


As we can see, the last example is incorrect, this is because emojis and certain character types fall our of our ASCII range of 0 to 127. We will change the criteria to three or more characters that fall out of the 0 to 127 range. 

In [15]:
def in_english(string):
    non_ASCII = 0 
    for character in string:
        if ord(character) > 127:
           non_ASCII += 1
    if non_ASCII > 3:
        return False
    else:
        return True
    
print(in_english('Instachat 😜'))

True


In [16]:
ios_english = []
android_english =[]
for app in ios_data:
    name = app[1]
    if in_english(name):
        ios_english.append(app)
        
for app in android_clean:
    name = app[0]
    if in_english(name):
        android_english.append(app)
        


print(len(ios_english))
print(len(android_english))

6183
9614


We've now removed non-english apps and the duplicate apps, next we need to isolate apps that are free.

In [28]:
free_android = []
free_ios = []

for app in android_english:
    price = app[6]
    if price == 'Free':
        free_android.append(app)

for app in ios_english:
    price = app[4]
    if price == '0.0':
        free_ios.append(app)
        
print(len(free_android))
print(len(free_ios))

8863
3222


To minimize risk and overhead, we initally build a minimal Android version of the app then add it to Google Play. If the feedback is good, we develop it further and if profitable after six months it will also be added to App Store. Thus we need to get an idea of an App that will work in both markets and looking at the genres that dominate these stores will help. 

In the App store we will look at the prime_genre column and the Genres and Category columns of the Google Play data set. We will do this by building a frequency table.

In [41]:
def freq_table(dataset, index):
    dictionary = {}
    dictionary_percentages = {}
    total = 0 
    for row in dataset:
        total += 1 
        column = row[index]
        if column in dictionary:
            dictionary[column] += 1
        else:
            dictionary[column] = 1 
            
    for key in dictionary:
        percentage = (dictionary[key]/total) * 100
        dictionary_percentages[key] = percentage
    return  dictionary_percentages

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
        
            


Examining the Prime Genre category in the App Store:

In [42]:
display_table(free_ios, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


The App Store is dominated by Games, which make up more than half of the total. Other popular apps are Entertainment, Photo and video apps. 

In [45]:
display_table(android_english, -4)

Tools : 8.602038693571874
Entertainment : 5.793634283336801
Education : 5.231953401289786
Business : 4.358227584772207
Medical : 4.108591637195756
Personalization : 3.900561680882047
Productivity : 3.879758685250676
Lifestyle : 3.775743707093822
Finance : 3.588516746411483
Sports : 3.442895776991887
Communication : 3.2660703141252343
Action : 3.110047846889952
Health & Fitness : 2.995631370917412
Photography : 2.9124193883919283
News & Magazines : 2.600374453921365
Social : 2.485957977948825
Travel & Local : 2.26752652381943
Books & Reference : 2.26752652381943
Shopping : 2.090701060952777
Simulation : 1.9762845849802373
Arcade : 1.9138755980861244
Dating : 1.768254628666528
Casual : 1.7162471395881007
Video Players & Editors : 1.674641148325359
Maps & Navigation : 1.3417932182234242
Puzzle : 1.2377782400665696
Food & Drink : 1.1649677553567712
Role Playing : 1.0817557728312877
Strategy : 0.9777407946744331
Racing : 0.9465363012273768
Libraries & Demo : 0.8737258165175785
Auto & Vehicl

Google store is  dominated by Tools, Education, Business etc apps, which are less "fun" and more practical than the top apps in the Apple Store

In [71]:
freq_table(ios_english,-5)

for genre in freq_table(ios_english, -5):
    total = 0
    len_genre = 0
    for app in ios_english:
        genre_app = app[-5]
        if genre_app == genre:
            user_ratings = float(app[5])
            total += user_ratings
            len_genre += 1 


    avg_user_ratings = total/len_genre
    print(genre, ':', avg_user_ratings)

Social Networking : 60253.84920634921
Photo & Video : 14688.715542521993
Games : 15586.759433962265
Music : 29047.109489051094
Reference : 27037.188679245282
Health & Fitness : 10802.157575757576
Weather : 23145.246376811596
Utilities : 7927.525821596244
Travel : 19030.183333333334
Shopping : 26635.011764705883
News : 16980.315789473683
Navigation : 19370.821428571428
Lifestyle : 8930.373737373737
Entertainment : 8862.409799554565
Food & Drink : 19934.386363636364
Sports : 15350.913461538461
Book : 10359.2
Finance : 23353.530612244896
Education : 2472.278048780488
Productivity : 8508.089285714286
Business : 5149.320754716981
Catalogs : 3465.0
Medical : 648.952380952381


In [72]:
for app in ios_english:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
MotionX GPS : 14970
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
Gaia GPS Classic : 2429
Plane Finder - Flight Tracker : 1438
iMaps+ for Google Maps ™ and Street View ™ : Transit and Offline Contacts : 1225
NAVIGON Europe : 927
Localscope - Find places and people around you : 868
Ski Tracks : 829
TRANSPORT MODS for MINECRAFT Pc EDITION : 754
Pocket Earth PRO Offline Maps & Travel Guides : 748
Ship Finder : 624
Boating USA : 342
Maps 3D PRO - GPS for Bike, Hike, Ski & Outdoor : 280
Cachly - Simple and powerful Geocaching for iPhone : 263
ImmobilienScout24: Real Estate Search in Germany : 187
The JMU Bus App : 35
Avertinoo : 32
iStellar : 30
mySTATE - State College : 26
Road watcher: dash camera, car video recorder. : 10
Streets – Street View Browser : 10
Railway Route Search : 5
parkOmator – for Apple Watch meter expiration timer, notifications & GPS navigator t

As we can see from the above, although navigation has a high number of average user ratings, it is skewed by big apps Waze and Google Maps. This is similar for Social Media Apps and Music Apps. We want to find a genre where the average user rating is more distrbuited.


In [73]:
for app in ios_english:
    if app[-5] == 'Health & Fitness':
        print(app[1], ':', app[5])

Calorie Counter & Diet Tracker by MyFitnessPal : 507706
Lose It! – Weight Loss Program and Calorie Counter : 373835
Weight Watchers : 136833
Sleep Cycle alarm clock : 104539
Fitbit : 90496
Period Tracker Lite : 53620
Nike+ Training Club - Workouts & Fitness Plans : 33969
White Noise : 33426
Yoga Studio : 28439
Plant Nanny - Water Reminder with Cute Plants : 27421
Runtastic PRO Running, Jogging and Fitness Tracker : 17667
Sworkit - Custom Workouts for Exercise & Fitness : 16819
Map My Run+ - GPS Running & Workout Tracker : 16434
MyNetDiary PRO - Calorie Counter and Food Diary : 15188
Sleep Pillow Sounds: white noise, rain, ocean, fan : 15116
Clue Period Tracker: Period & Ovulation Tracker : 13436
Period Tracker Deluxe : 13350
Headspace : 12819
Fooducate - Lose Weight, Eat Healthy,Get Motivated : 11875
Fitness Buddy+ Gym Workout Exercise, Home Trainer : 10386
5K Runner: 0 to 5K Run Trainer. Couch potato to 5K : 10368
Runtastic Running, Jogging and Walking Tracker : 10298
Instant Heart Ra

Not including myfitnesspal, 'Health & Fitness' might be a good genre to go to with several different apps dominating the genre. More "fun" apps like social media are heavily saturated, so building a new fitness app with gyms being closed has potential.

For the Google Market store, we have the number of installs. However they are not accurate we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [74]:
display_table(android_english, 5) # installs

1,000,000+ : 14.70771791137924
100,000+ : 11.504056584148117
10,000+ : 10.619929269814854
10,000,000+ : 9.746203453297275
1,000+ : 9.153318077803203
100+ : 7.322654462242563
5,000,000+ : 6.292906178489703
500,000+ : 5.242354899105471
5,000+ : 4.836696484293738
50,000+ : 4.815893488662367
10+ : 3.994175161223216
500+ : 3.4116912835448305
50,000,000+ : 2.1219055543998335
50+ : 2.1219055543998335
100,000,000+ : 1.9658830871645518
5+ : 0.8529228208862076
1+ : 0.6864988558352403
500,000,000+ : 0.249635947576451
1,000,000,000+ : 0.20802995631370919
0+ : 0.13521947160391096
0 : 0.010401497815685458


In [80]:
genres_android = freq_table(android_english, 1)

for genres in genres_android:
    total = 0
    len_genres = 0
    for app in android_english:
        genres_app = app[1]
        if genres_app == genres:            
            installs = app[5]
            installs = installs.replace(',', '')
            installs = installs.replace('+', '')
            total += float(installs)
            len_genres += 1
    avg_no_of_installs = total / len_genres
    print(genres, ':', avg_no_of_installs)

ART_AND_DESIGN : 1887285.0
AUTO_AND_VEHICLES : 632501.3214285715
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 7641777.871559633
BUSINESS : 1663758.627684964
COMICS : 817657.2727272727
COMMUNICATION : 35153714.17515924
DATING : 828971.2176470588
EDUCATION : 1782566.0377358492
ENTERTAINMENT : 11375402.298850575
EVENTS : 249580.640625
FINANCE : 1319851.4028985507
FOOD_AND_DRINK : 1891060.2767857143
HEALTH_AND_FITNESS : 3972300.388888889
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 630903.6904761905
LIFESTYLE : 1369954.7774725275
GAME : 14256217.600635594
FAMILY : 3345018.516684607
MEDICAL : 96944.49873417722
SOCIAL : 22961790.384937238
SHOPPING : 6966908.880597015
PHOTOGRAPHY : 16636241.267857144
SPORTS : 3373767.6861538463
TRAVEL_AND_LOCAL : 13218662.767123288
TOOLS : 9785955.211352658
PERSONALIZATION : 4086652.4853333333
PRODUCTIVITY : 15530942.008042896
PARENTING : 525351.8333333334
WEATHER : 4570892.658227848
VIDEO_PLAYERS : 24121489.079754602
NEWS_AND_MAGAZINES : 947

In [82]:
for app in android_english:
    if app[1] == 'HEALTH_AND_FITNESS':
        print(app[0], ':', app[5])

Step Counter - Calorie Counter : 500,000+
Lose Belly Fat in 30 Days - Flat Stomach : 5,000,000+
Pedometer - Step Counter Free & Calorie Burner : 1,000,000+
Six Pack in 30 Days - Abs Workout : 10,000,000+
Lose Weight in 30 Days : 10,000,000+
Pedometer : 10,000,000+
LG Health : 10,000,000+
Step Counter - Pedometer Free & Calorie Counter : 10,000,000+
Pedometer, Step Counter & Weight Loss Tracker App : 10,000,000+
Sportractive GPS Running Cycling Distance Tracker : 1,000,000+
30 Day Fitness Challenge - Workout at Home : 10,000,000+
Home Workout for Men - Bodybuilding : 1,000,000+
Fat Burning Workout - Home Weight lose : 100,000+
Buttocks and Abdomen : 500,000+
Walking for Weight Loss - Walk Tracker : 100,000+
Running & Jogging : 500,000+
Sleep Sounds : 1,000,000+
Fitbit : 10,000,000+
Lose Belly Fat-Home Abs Fitness Workout : 50,000+
Cycling - Bike Tracker : 500,000+
Abs Training-Burn belly fat : 100,000+
Calorie Counter - EasyFit free : 1,000,000+
Aunjai i lert u : 500,000+
Garmin Connect

The market doesn't seem to be dominated by a few apps in the Google play store either.  Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

Conclusions

We analysed data about the App Store and Google Play mobile apps with the aim of recommending an app profile that can be profitable for both markets.

In the mid-section of the Health and Fitness market it is dominated by apps that show different workouts and at the top end by calorie counters. It's perhaps best then to develop an app that provides users with different workout routines and somewhere where they can also track their nutrition. 
